### DISCUSSION TITLE
Is there a way to automatically upgrade outdated bottles?

### DISCUSSION THREAD

**Discussion by ywwry66** [Homebrew/discussions#308] [Everyday usage] [answered]
Today, I tried to install some R package, which complained that it couldn't find the path `/usr/local/opt/gcc/lib/gcc/10/gcc/x86_64-apple-darwin20/10.2.0`. I then figured out this was because the `gcc` I installed was bottled under Catalina, which is shipped with the directory `x86_64-apple-darwin19` instead of `x86_64-apple-darwin20`. This can be easily solved by `brew reinstall gcc`, which upgrades to the latest Big Sur bottle.

However, this process is a pain without automation. Of course I can reinstall all formulae when I upgrade to each year's major release of macOS (and I will have to wait until all the new bottles are built). But this seems to be an overkill. Anyway, I was wondering if there is some kind of best practice regarding this issue. Thank you.

**Comment by gromgit** [comment, marked as answer]
I just upgraded an old box from Mojave to Catalina, and chose the "nuclear option":
```sh
xcode-select --install  # upgrade the existing CLT, else non-bottled formulae will fail to reinstall
brew update
brew ls | xargs brew reinstall 2>&1 | tee /tmp/reinstall.log
```
If you want to ensure that only bottles are used, substitute `brew reinstall --force-bottle` above, then look in `/tmp/reinstall.log` to see what failed to update, then decide how to proceed with each failure.

**Reply by ywwry66** [reply]
Thanks. This would work if most formulae are already bottled. I guess I'd better hold off from upgrading to the latest OS until most bottles are built, so then this "nuclear option" can apply.

Another thing I was thinking is, should `brew upgrade` also check and upgrade a formula to the latest bottle version? This will prevent the issue I encountered at the cost of a little bit of efficiency.

**Reply by gromgit** [reply]
> should `brew upgrade` also check and upgrade a formula to the latest bottle version?

I can't speak for the maintainers, but I'd personally prefer if a routine `brew upgrade` didn't start pulling down countless gigabytes to update almost 500 kegs, just because it happened right after a OS upgrade:
```
$ ls -d /usr/local/opt/* | wc -l
456
```
In fact, the only reason I responded to this discussion was because I went "hmmm, I wonder how I would go about it?" and decided to test it. Otherwise, I would've left it alone (most macOS binaries don't break after an OS upgrade), only done a `brew reinstall` on specific packages when they broke, and let `brew upgrade` refresh the rest when their versions get bumped.

**Reply by carlocab** [reply]
>most macOS binaries don't break after an OS upgrade

I think this is right. Leading exceptions are compilers like `gcc` and `llvm`, plus a handful of other formulae that hard-code the SDK path into their binary artefacts.

**Reply by ywwry66** [reply]
Thanks, it is nice to have these few exceptions in mind.
